[SPARK-45861][PYTHON][DOCS] Add user guide for dataframe creation (#43897)
apache/spark: python/docs/source/user_guide/dataframe_creation.rst

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

==================

> DataFrame creation

Reviewer: Maybe shall we capitalizing the first word of the title/sub-title for consistency with other documents?

e.g. DataFrame Creation / Basic Data Structures  ...

==================

.. currentmodule:: pyspark.sql

Basic data structures
---------------------

> Pyspark provides an important class for handling data:

Reviewer: Pyspark -> PySpark

1. :class:`DataFrame`: a distributed collection of data grouped into named columns.

Creating through `createDataFrame`
----------------------------------

> A PySpark :class:`DataFrame` can be created via :meth:`SparkSession.createDataFrame` typically by passing
> a list of lists, tuples, dictionaries and :class:`Row`, a pandas :class:`pandas.DataFrame`
> and an :class:`pyspark.RDD` consisting of such a list.

Reviewer: Maybe do we want to mention  :class:\`numpy.ndarray\`  here as well? e.g.

```suggestion
A PySpark :class:`DataFrame` can be created via :meth:`SparkSession.createDataFrame` typically by passing a list of lists, tuples, dictionaries, :class:`Row`, a pandas :class:`pandas.DataFrame`, a NumPy :class:\`numpy.ndarray\` and an :class:`pyspark.RDD`.
```

> :meth:`SparkSession.createDataFrame` takes the `schema` argument to specify the schema of the DataFrame.

Reviewer: DataFrame -> :class:\`DataFrame\`

When it is omitted, PySpark infers the corresponding schema by taking a sample from the data.

> Creating a PySpark :class:`DataFrame` from a list of lists

Reviewer 2: shall we also add examples for `schema` in `createDataFrame`?

Author: Okay

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    df = spark.createDataFrame([['Alice', 1], ['Bob', 5]])

>     df

Reviewer 2: what about using `df.show()` in the examples instead to show the data?

Author: Okay

    df

DataFrame[age: bigint, name: string]

Creating a PySpark :class:`DataFrame` from a list of :class:`Row`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    from pyspark.sql import Row
    Person = Row('name', 'age')
    df = spark.createDataFrame([Person("Alice", 1), Person("Bob", 5)])
    df

DataFrame[name: string, age: bigint]

Creating a PySpark :class:`DataFrame` from a :class:`pandas.DataFrame`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    import pandas as pd
    df = spark.createDataFrame(pd.DataFrame([[1, 2]]))
    df

DataFrame[0: bigint, 1: bigint]

Creating a PySpark :class:`DataFrame` from a :class:`numpy.ndarray`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    import numpy as np
    import pandas as pd
    df = spark.createDataFrame(pd.DataFrame(data=np.array([[1, 2], [3, 4]]), columns=['a', 'b']))
    df

DataFrame[a: bigint, b: bigint]

Creating a PySpark :class:`DataFrame` from an :class:`pyspark.RDD`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> .. code-block:: python

Reviewer 2: I think we'd better not mention RDD since it is not supported on connect

Reviewer: > I think we'd better not mention RDD since it is not supported on connect

+1, or at least we should add a note.

..  Licensed to the Apache Software Foundation (ASF) under one
    or more contributor license agreements.  See the NOTICE file
    distributed with this work for additional information
    regarding copyright ownership.  The ASF licenses this file
    to you under the Apache License, Version 2.0 (the
    "License"); you may not use this file except in compliance
    with the License.  You may obtain a copy of the License at

..    http://www.apache.org/licenses/LICENSE-2.0

..  Unless required by applicable law or agreed to in writing,
    software distributed under the License is distributed on an
    "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
    KIND, either express or implied.  See the License for the
    specific language governing permissions and limitations
    under the License.

==================
DataFrame Creation
==================

.. currentmodule:: pyspark.sql

> Basic Data Structures

Reviewer 3: This isn't really useful here as it includes one class alone. Should probably just remove them

Reviewer 3: Or, include the methods being used here. e.g., `` :meth:`SparkSession.createDataFrame ``, `` :meth:`DataFrameReader.load` ``, etc.

:meth:`SparkSession.createDataFrame` takes the `schema` argument to specify the schema of the :class:`DataFrame`.
When it is omitted, PySpark infers the corresponding schema by taking a sample from the data.

Creating a PySpark :class:`DataFrame` from a list of lists
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> df = spark.createDataFrame([['Alice', 1], ['Bob', 5]])
    >>> df.show()
    +-----+---+
    |   _1| _2|
    +-----+---+
    |Alice|  1|
    |  Bob|  5|
    +-----+---+

Creating a PySpark :class:`DataFrame` from a list of tuples
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> df = spark.createDataFrame([('Alice', 1), ('Bob', 5)])
    >>> df.show()
    +-----+---+
    |   _1| _2|
    +-----+---+
    |Alice|  1|
    |  Bob|  5|
    +-----+---+

Creating a PySpark :class:`DataFrame` with the explicit schema specified
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> from pyspark.sql.types import *
    >>> schema = StructType([StructField("name", StringType(), True),
    ...     StructField("age", IntegerType(), True)])

>     >>> df = spark.createDataFrame([('Alice', 1), ('Bob', 5)], schema)

Reviewer 3: We should probably show an example with a string DDL schema schema

Creating a PySpark :class:`DataFrame` from a list of :class:`Row`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> from pyspark.sql import Row
    >>> Person = Row('name', 'age')
    >>> df = spark.createDataFrame([Person("Alice", 1), Person("Bob", 5)])
    >>> df.show()
    +-----+---+
    | name|age|
    +-----+---+
    |Alice|  1|
    |  Bob|  5|
    +-----+---+

Creating a PySpark :class:`DataFrame` from a :class:`pandas.DataFrame`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> import pandas as pd
    >>> df = spark.createDataFrame(pd.DataFrame([[1, 2]]))
    >>> df.show()
    +---+---+
    |  0|  1|
    +---+---+
    |  1|  2|
    +---+---+

Creating a PySpark :class:`DataFrame` from a :class:`numpy.ndarray`
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> import numpy as np
    >>> import pandas as pd
    >>> df = spark.createDataFrame(pd.DataFrame(data=np.array([[1, 2], [3, 4]]),
    ...     columns=['a', 'b']))
    >>> df.show()
    +---+---+
    |  a|  b|
    +---+---+
    |  1|  2|
    |  3|  4|
    +---+---+

> Creating Through `read.format(...).load(...)`

Reviewer 3: Preposition starts with a lower case.

```suggestion
Creating through `read.format(...).load(...)`
```

---------------------------------------------

Creating a PySpark :class:`DataFrame` by reading existing **json** format file data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> df = spark.read.format("json").load("python/test_support/sql/people.json")
    >>> df.show()
    +----+-------+
    | age|   name|
    +----+-------+
    |NULL|Michael|
    |  30|   Andy|
    |  19| Justin|
    +----+-------+

Creating a PySpark :class:`DataFrame` by reading existing **csv** format file data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

>     >>> df = spark.read.format("csv").option("header", "true").
>     ...     load("python/test_support/sql/people.csv")

Reviewer 3: This won't work because of the indentation.
```suggestion
    >>> df = spark.read.format("csv").option("header", "true").load(
    ...     "python/test_support/sql/people.csv")
```

    >>> df.show()
    +----+-------+
    | age|   name|
    +----+-------+
    |NULL|Michael|
    |  30|   Andy|
    |  19| Justin|
    +----+-------+

Creating a PySpark :class:`DataFrame` by reading existing **parquet** format file data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

>     >>> df = spark.read.format("parquet").load("python/test_support/sql/people.parquet")

Reviewer 3: I think we should make the example self-contained with including `spark.range(10).write("parquet")` with properly adding comments.

    >>> df.show()
    +----+-------+
    | age|   name|
    +----+-------+
    |NULL|Michael|
    |  30|   Andy|
    |  19| Justin|
    +----+-------+

Creating a PySpark :class:`DataFrame` by reading existing **orc** format file data
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

.. code-block:: python

    >>> df = spark.read.format("parquet").load("python/test_support/sql/people.orc")
    >>> df.show()
    +----+-------+
    | age|   name|
    +----+-------+
    |NULL|Michael|
    |  30|   Andy|
    |  19| Justin|
    +----+-------+

Creating a PySpark :class:`DataFrame` by reading data from other databases using **JDBC**
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 

Reviewer 3: I think we should really have a self-contained example, e.g. https://sparkbyexamples.com/pyspark/pyspark-read-and-write-mysql-database-table/. We could have a separate page for this, and link it here. Could be done separately in another PR.

Author: Okay, Let me temporarily remove this section from the document. 
I have already created a new jira, and when I start this jira https://issues.apache.org/jira/browse/SPARK-46088, 
I will add this section back and link it.
